WK6_DAY3_DC - 9 JUILLET 2025

In [ ]:
# Classify Handwritten Digits with CNNs

# Load the MNIST dataset:

# Installation de TensorFlow (si nécessaire) pour pouvoir charger le jeu de données MNIST
!pip install -q tensorflow

# Importation après installation
import numpy as np
import matplotlib.pyplot as plt
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten, Conv2D, MaxPooling2D
from tensorflow.keras.utils import to_categorical

In [ ]:
# Étape 2 : Chargement du jeu de données MNIST

from tensorflow.keras.datasets import mnist

# Charger les données
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# Afficher les dimensions
print("x_train shape :", x_train.shape)
print("y_train shape :", y_train.shape)
print("x_test shape  :", x_test.shape)
print("y_test shape  :", y_test.shape)

x_train shape : (60000, 28, 28)
y_train shape : (60000,)
x_test shape  : (10000, 28, 28)
y_test shape  : (10000,)


In [ ]:
# Étape 1 : Flatten
x_train_flat = x_train.reshape(x_train.shape[0], 28*28)
x_test_flat = x_test.reshape(x_test.shape[0], 28*28)

# Étape 2 : Normalisation
x_train_flat = x_train_flat / 255.0
x_test_flat = x_test_flat / 255.0

# Étape 3 : One-hot encoding
from tensorflow.keras.utils import to_categorical
y_train_cat = to_categorical(y_train, num_classes=10)
y_test_cat = to_categorical(y_test, num_classes=10)

In [ ]:
# Build and train a Fully Connected Neural Network:
    # Create a Sequential model.
    # Add Dense layers with appropriate activation functions (e.g., ReLU and softmax).
    # Compile the model with an optimizer (e.g., Adam), loss function (e.g., categorical cross-entropy), and metrics (e.g., accuracy).
    # Train the model using model.fit() and evaluate its performance.#

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.utils import to_categorical

# Créer et compiler le modèle
model_fcn = Sequential([
    Dense(128, activation='relu', input_shape=(784,)),
    Dense(64, activation='relu'),
    Dense(10, activation='softmax')
])

model_fcn.compile(optimizer='adam',
                  loss='categorical_crossentropy',
                  metrics=['accuracy'])

# Entraîner
history_fcn = model_fcn.fit(x_train_flat, y_train_cat,
                            epochs=10,
                            batch_size=32,
                            validation_split=0.1,
                            verbose=1)

# Évaluer
test_loss, test_accuracy = model_fcn.evaluate(x_test_flat, y_test_cat, verbose=0)
print(f"Test Accuracy: {test_accuracy:.4f}")

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.8671 - loss: 0.4557 - val_accuracy: 0.9697 - val_loss: 0.1107
Epoch 2/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 11s 4ms/step - accuracy: 0.9649 - loss: 0.1153 - val_accuracy: 0.9685 - val_loss: 0.1071
Epoch 3/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9775 - loss: 0.0733 - val_accuracy: 0.9690 - val_loss: 0.0971
Epoch 4/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9823 - loss: 0.0558 - val_accuracy: 0.9755 - val_loss: 0.0784
Epoch 5/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9879 - loss: 0.0401 - val_accuracy: 0.9768 - val_loss: 0.0859
Epoch 6/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step - accuracy: 0.9907 - loss: 0.0287 - val_accuracy: 0.9775 - val_loss: 0.0791
Epoch 7/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9915 - loss: 0.0255 - val_accuracy: 0.9787 - val_loss: 0.0915
Epoch 8/10
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step - accuracy: 0.9934 - loss: 0.0195 -

In [3]:
# Preprocess the data for a Convolutional Neural Network:
    # Reshape the input data to the shape expected by a Conv2D layer (e.g., (60000, 28, 28, 1)).
    # Normalize the pixel values by dividing by 255.
    # One-hot encode the target labels using keras.utils.np_utils.to_categorical().

from tensorflow.keras.utils import to_categorical

# Rechargement des données MNIST pour s'assurer que tout est bien défini
from tensorflow.keras.datasets import mnist
(x_train_cnn, y_train_cnn), (x_test_cnn, y_test_cnn) = mnist.load_data()

# 1. Reshape : ajout d'une dimension "canal" (grayscale = 1 canal)
x_train_cnn = x_train_cnn.reshape(-1, 28, 28, 1)
x_test_cnn = x_test_cnn.reshape(-1, 28, 28, 1)

# 2. Normalisation des pixels (valeurs entre 0 et 1)
x_train_cnn = x_train_cnn / 255.0
x_test_cnn = x_test_cnn / 255.0

# 3. One-hot encoding des étiquettes
y_train_cnn_cat = to_categorical(y_train_cnn, num_classes=10)
y_test_cnn_cat = to_categorical(y_test_cnn, num_classes=10)

# Vérification des shapes
x_train_cnn.shape, y_train_cnn_cat.shape, x_test_cnn.shape, y_test_cnn_cat.shape

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


((60000, 28, 28, 1), (60000, 10), (10000, 28, 28, 1), (10000, 10))

In [4]:
# Build and train a Convolutional Neural Network:
    # Create a Sequential model.
    # Add Conv2D and MaxPool2D layers.
    # Add a Flatten layer.
    # Add Dense layers with appropriate activation functions.
    # Compile and train the model, similar to the Fully Connected Neural Network.

# Installation de TensorFlow pour permettre la construction et l'entraînement du modèle CNN
!pip install -q tensorflow

# Redémarrage automatique du noyau à faire manuellement si nécessaire dans certains environnements
# Rechargement après installation
import tensorflow as tf
from tensorflow.keras import layers, models


# Chargement des données MNIST
(x_train_cnn, y_train_cnn), (x_test_cnn, y_test_cnn) = mnist.load_data()

# Prétraitement
x_train_cnn = x_train_cnn.reshape(-1, 28, 28, 1).astype("float32") / 255.0
x_test_cnn = x_test_cnn.reshape(-1, 28, 28, 1).astype("float32") / 255.0
y_train_cnn_cat = to_categorical(y_train_cnn, num_classes=10)
y_test_cnn_cat = to_categorical(y_test_cnn, num_classes=10)

# Création du modèle CNN
model_cnn = models.Sequential([
    layers.Conv2D(32, kernel_size=(3,3), activation='relu', input_shape=(28,28,1)),
    layers.MaxPooling2D(pool_size=(2,2)),
    layers.Conv2D(64, kernel_size=(3,3), activation='relu'),
    layers.MaxPooling2D(pool_size=(2,2)),
    layers.Flatten(),
    layers.Dense(128, activation='relu'),
    layers.Dense(10, activation='softmax')  # 10 classes pour les chiffres de 0 à 9
])

# Compilation du modèle
model_cnn.compile(optimizer='adam',
                  loss='categorical_crossentropy',
                  metrics=['accuracy'])

# Entraînement du modèle
history_cnn = model_cnn.fit(
    x_train_cnn, y_train_cnn_cat,
    epochs=10,
    batch_size=64,
    validation_split=0.1,
    verbose=1
)

# Évaluation sur les données de test
test_loss_cnn, test_accuracy_cnn = model_cnn.evaluate(x_test_cnn, y_test_cnn_cat, verbose=0)
test_loss_cnn, test_accuracy_cnn

/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/10
844/844 ━━━━━━━━━━━━━━━━━━━━ 41s 47ms/step - accuracy: 0.8786 - loss: 0.3983 - val_accuracy: 0.9847 - val_loss: 0.0521
Epoch 2/10
844/844 ━━━━━━━━━━━━━━━━━━━━ 39s 46ms/step - accuracy: 0.9836 - loss: 0.0548 - val_accuracy: 0.9893 - val_loss: 0.0379
Epoch 3/10
844/844 ━━━━━━━━━━━━━━━━━━━━ 41s 46ms/step - accuracy: 0.9891 - loss: 0.0354 - val_accuracy: 0.9855 - val_loss: 0.0443
Epoch 4/10
844/844 ━━━━━━━━━━━━━━━━━━━━ 41s 46ms/step - accuracy: 0.9909 - loss: 0.0266 - val_accuracy: 0.9920 - val_loss: 0.0291
Epoch 5/10
844/844 ━━━━━━━━━━━━━━━━━━━━ 41s 46ms/step - accuracy: 0.9946 - loss: 0.0174 - val_accuracy: 0.9910 - val_loss: 0.0323
Epoch 6/10
844/844 ━━━━━━━━━━━━━━━━━━━━ 41s 46ms/step - accuracy: 0.9950 - loss: 0.0145 - val_accuracy: 0.9882 - val_loss: 0.0423
Epoch 7/10
844/844 ━━━━━━━━━━━━━━━━━━━━ 41s 46ms/step - accuracy: 0.9958 - loss: 0.0124 - val_accuracy: 0.9907 - val_loss: 0.0351
Epoch 8/10
844/844 ━━━━━━━━━━━━━━━━━━━━ 41s 46ms/step - accuracy: 0.9966 - loss: 0.0096 - 

(0.03318173065781593, 0.9908000230789185)

# Compare the performance:
    # Analyze the accuracy of both models.
    # Observe the difference between the Fully connected model and the CNN model.

Comparaison des performances entre modèles Fully Connected (Dense) et Convolutional Neural Network (CNN) construits sur le dataset MNIST de chiffres manuscrits :

1. Analyze the accuracy of both models
Accuracy des deux modèles sur les données de test
    * Fully Connected Network	Environ 0.97 (97%)
    * CNN (Convolutional)	Environ 0.99 (99%)

2. Observe the difference between the Fully connected model and the CNN model.
La performance des 2 modèles est de très bon niveau; toutefois, le CNN est le plus performant des 2 modèles.

Conclusion :
 - Le réseau dense fonctionne bien si on écrase les images (flatten), mais il perd la notion de voisinage de pixels.
 - Le réseau convolutionnel est spécialement conçu pour traiter des images : il "regarde" les petits morceaux de l’image, détecte des motifs, et les combine pour faire de meilleures prédictions.